src/librustdoc/html/render/cache.rs

```diff
@@ -219,18 +225,139 @@ fn get_index_type_name(clean_type: &clean::Type, accept_generic: bool) -> Option
     }
 }
 
-fn get_generics(clean_type: &clean::Type, cache: &Option<&Cache>) -> Option<Vec<Generic>> {
+fn get_generics(clean_type: &clean::Type, cache: &Cache) -> Option<Vec<Generic>> {
     clean_type.generics().and_then(|types| {
         let r = types
             .iter()
             .filter_map(|t| {
                 get_index_type_name(t, false).map(|name| Generic {
                     name: name.as_str().to_ascii_lowercase(),
-                    defid: cache.map_or_else(|| t.def_id(), |cache| t.def_id_full(cache)),
+                    defid: t.def_id_full(cache),
                     idx: None,
                 })
             })
             .collect::<Vec<_>>();
         if r.is_empty() { None } else { Some(r) }
     })
 }
+
+/// The point of this function is to replace bounds with types.
+///
+/// i.e. `[T, U]` when you have the following bounds: `T: Display, U: Option<T>` will return
+/// `[Display, Option]` (we just returns the list of the types, we don't care about the
+/// wrapped types in here).
+crate fn get_real_types<'tcx>(
+    generics: &Generics,
+    arg: &Type,
+    tcx: TyCtxt<'tcx>,
+    recurse: i32,
+    cache: &Cache,
+    res: &mut FxHashSet<(Type, TypeKind)>,
+) -> usize {
+    if recurse >= 10 {
+        // FIXME: remove this whole recurse thing when the recursion bug is fixed
+        return 0;
+    }
+    let mut nb_added = 0;
+
+    if arg.is_full_generic() {
+        let arg_s = Symbol::intern(&arg.print(cache).to_string());
```

Comment:
Except that the `cache`, even if useless now is useful since https://github.com/rust-lang/rust/pull/81557/files#diff-3bcc7ccbdf24afcb04155aacd071daa66d96350b30c5a2991e2f78b943699863. I wrote the full cache rewrite PR by letting the `cache` here knowing that we would need it for the primitive search. ;)

However, the improvement you suggested is very nice, so still putting it in!

Reply:
Hold on, I need to check if we need to turn them into `def_id_full` here in the other PR.

Reply:
Ok, it needs to have the same logic as in #81557. I'll cherry-pick some commits from there to ensure that we're all good in here.

Reply:
Hum, conflicts, better wait for #81557 to be merged before then to be able to work with peaceful mind in here.

Reply:
> Except that the `cache`, even if useless now is useful since https://github.com/rust-lang/rust/pull/81557/files#diff-3bcc7ccbdf24afcb04155aacd071daa66d96350b30c5a2991e2f78b943699863.

That change didn't do anything. The function arguments in the search index don't use `DefId`s. They're added to the `RenderType` struct but never read. The reason the `dead_code` lint doesn't catch that is because of the `#[derive(Debug)]`. In fact the only field in `RenderType` that's actually used is `name` so unless there are future plans for `RenderType` it can and should be removed.

Reply:
You're absolutely right. I was confused becase of the usage of the `cache` to get the primitive types but it doesn't make sense here. Removing it.